firstpartydata/first_party_data_test.go

```diff
@@ -0,0 +1,836 @@
+		for _, specFile := range specFiles {
+			fileName := "./tests/resolvefpd/" + specFile.Name()
+
+			fpdFile, err := loadFpdFile(fileName)
+			if err != nil {
+				t.Errorf("Unable to load file: %s", fileName)
+			}
+
+			var inputReq openrtb2.BidRequest
+			err = json.Unmarshal(fpdFile.InputRequestData, &inputReq)
+			if err != nil {
+				t.Errorf("Unable to unmarshal input request: %s", fileName)
+			}
+
+			var inputReqCopy openrtb2.BidRequest
+			err = json.Unmarshal(fpdFile.InputRequestData, &inputReqCopy)
+			if err != nil {
+				t.Errorf("Unable to unmarshal input request: %s", fileName)
+			}
+
+			var outputReq openrtb2.BidRequest
+			err = json.Unmarshal(fpdFile.OutputRequestData, &outputReq)
+			if err != nil {
+				t.Errorf("Unable to unmarshal output request: %s", fileName)
+			}
+
+			reqExtFPD := make(map[string][]byte, 3)
+			reqExtFPD["site"] = fpdFile.FirstPartyData["site"]
+			reqExtFPD["app"] = fpdFile.FirstPartyData["app"]
+			reqExtFPD["user"] = fpdFile.FirstPartyData["user"]
+
+			reqFPD := make(map[string][]openrtb2.Data, 3)
+
+			reqFPDSiteContentData := fpdFile.FirstPartyData[siteContentDataKey]
+			if len(reqFPDSiteContentData) > 0 {
+				var siteConData []openrtb2.Data
+				err = json.Unmarshal(reqFPDSiteContentData, &siteConData)
+				if err != nil {
+					t.Errorf("Unable to unmarshal site.content.data: %s", fileName)
+				}
+				reqFPD[siteContentDataKey] = siteConData
+			}
+
+			reqFPDAppContentData := fpdFile.FirstPartyData[appContentDataKey]
+			if len(reqFPDAppContentData) > 0 {
+				var appConData []openrtb2.Data
+				err = json.Unmarshal(reqFPDAppContentData, &appConData)
+				if err != nil {
+					t.Errorf("Unable to unmarshal app.content.data: %s", fileName)
+				}
+				reqFPD[appContentDataKey] = appConData
+			}
+
+			reqFPDUserData := fpdFile.FirstPartyData[userDataKey]
+			if len(reqFPDUserData) > 0 {
+				var userData []openrtb2.Data
+				err = json.Unmarshal(reqFPDUserData, &userData)
+				if err != nil {
+					t.Errorf("Unable to unmarshal app.content.data: %s", fileName)
+				}
+				reqFPD[userDataKey] = userData
+			}
+			if fpdFile.BiddersFPD == nil {
+				fpdFile.BiddersFPD = make(map[openrtb_ext.BidderName]*openrtb_ext.ORTB2)
+				fpdFile.BiddersFPD["appnexus"] = &openrtb_ext.ORTB2{}
+			}
+
+			resultFPD, errL := ResolveFPD(&inputReq, fpdFile.BiddersFPD, reqExtFPD, reqFPD, []string{"appnexus"})
+
+			if len(errL) == 0 {
+				assert.Equal(t, inputReq, inputReqCopy, "Original request should not be modified")
+
+				bidderFPD := resultFPD["appnexus"]
+
+				if bidderFPD.Site != nil && len(bidderFPD.Site.Ext) > 0 {
+					resSiteExt := bidderFPD.Site.Ext
+					expectedSiteExt := outputReq.Site.Ext
+					bidderFPD.Site.Ext = nil
+					outputReq.Site.Ext = nil
+					assert.JSONEq(t, string(expectedSiteExt), string(resSiteExt), "site.ext is incorrect")
+
+					assert.Equal(t, outputReq.Site, bidderFPD.Site, "Site is incorrect")
+				}
+				if bidderFPD.App != nil && len(bidderFPD.App.Ext) > 0 {
```

Comment:
Again, if the app object in the result is nil but we actually expected an app object then this test will silently pass. Same for site and user

Reply:
Agree, modified it to check it with expected result, not actual